src/Client.php

```diff
@@ -76,4 +77,113 @@ public function iterate(string $method, array $arguments): iterable
+     *                         'title' => 'My File', // optional
+     *                         'alt_text' => 'Screen reader description', // optional
+     *                         'snippet_type' => 'php', // optional
+     *                     ]
+     * @param string $channelId The Slack channel ID to upload files to (e.g., 'C12345678').
+     * @param string|null $initialComment Optional comment to add with the upload.
+     *
+     * @return \JoliCode\Slack\Api\Model\FilesCompleteUploadExternalPostResponse200|\JoliCode\Slack\Api\Model\FilesCompleteUploadExternalPostResponsedefault|null
+     *
+     * @throws \RuntimeException If upload or Slack API interaction fails.
+     *
+     * @example
+     * $this->filesUploadV2([
+     *     ['path' => '/tmp/image.png', 'title' => 'Bot Logo', 'alt_text' => 'Slack Bot Logo'],
+     *     ['path' => '/tmp/script.js', 'snippet_type' => 'javascript']
+     * ], 'C12345678', 'Here are the uploaded files:');
+     */
+    public function filesUploadV2(array $files, string $channelId, ?string $initialComment = null)
+    {
+        $filesPayload = [];
+
+        foreach ($files as $file) {
+            $filePath = $file['path'];
+            $fileName = basename($filePath);
+            $title = $file['title'] ?? $fileName;
+            $altText = $file['alt_text'] ?? null;
+            $snippetType = $file['snippet_type'] ?? null;
+
+            $fileStream = Stream::create(fopen($filePath, 'r'));
+            $fileSize = $fileStream->getSize();
+
+            // Step 1: Get upload URL from Slack
+            $queryParams = array_filter(
+                [
+                    'filename' => $fileName,
+                    'length' => $fileSize,
+                    'alt_txt' => $altText,
+                    'snippet_type' => $snippetType,
+                ]
+            );
+
+            $uploadInfo = $this->filesGetUploadUrlExternal($queryParams);
+            $uploadUrl = $uploadInfo->getUploadUrl();
+            $fileId = $uploadInfo->getFileId();
+
+            if (!$uploadUrl || !$fileId) {
+                throw new \RuntimeException("Slack did not return upload URL or file ID for {$fileName}");
+            }
+
+            // Step 2: Upload file data using cURL
+            $this->uploadToSlackUrlWithCurl($uploadUrl, $filePath);
+
+            // Step 3: Build file metadata
+            $filesPayload[] = [
+                'id' => $fileId,
+                'title' => $title,
+            ];
+        }
+
+        // Step 4: Complete upload
+        return $this->filesCompleteUploadExternal(
+            array_filter(
+                [
+                    'files' => json_encode($filesPayload),
+                    'channel_id' => $channelId,
+                    'initial_comment' => $initialComment,
+                ]
+            )
+        );
+    }
+
+    private function uploadToSlackUrlWithCurl(string $uploadUrl, string $filePath): void
```

Comment:
It would be nice to use the http client user in the actual api client (Jane).

Would that be possible?

Reply:
That was discussed here too: https://github.com/jolicode/slack-php-api/pull/176

I like what you did yes :+1: much more closer to the actual API client that way. I will suggest a modification tho.